kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/V3ResourcesModule.java

```diff
@@ -30,5 +39,38 @@ protected void configure() {
     bindAsContract(ChunkedOutputFactory.class);
     bindAsContract(StreamingResponseFactory.class);
     bind(Clock.systemUTC()).to(Clock.class);
+    bindFactory(ProduceResponseExecutorServiceFactory.class)
+        .to(ExecutorService.class)
+        .in(Singleton.class);
+  }
+
+  private static final class ProduceResponseExecutorServiceFactory
+      implements Factory<ExecutorService> {
+
+    private final int produceResponseThreadPoolSize;
+
+    @Inject
+    ProduceResponseExecutorServiceFactory(
+        @ProduceResponseThreadPoolSizeConfig Integer produceExecutorThreadPoolSize) {
+      this.produceResponseThreadPoolSize = produceExecutorThreadPoolSize;
+    }
+
+    @Override
+    public ExecutorService provide() {
+      ThreadFactory namedThreadFactory =
+          new ThreadFactoryBuilder().setNameFormat("Produce-response-thread-%d").build();
+      return new ThreadPoolExecutor(
+          produceResponseThreadPoolSize,
+          produceResponseThreadPoolSize,
+          0L,
+          TimeUnit.MILLISECONDS,
+          new LinkedBlockingQueue<Runnable>(),
+          namedThreadFactory);
+    }
+
+    @Override
+    public void dispose(ExecutorService executorService) {
+      executorService.shutdown();
+    }
```

Comment:
not-a-comment: I still haven't learned how does HK2 behave for singletons - do you know when will this be called (also CC @rigelbm as he probably knows)?

Reply:
At the end of the instance life-cycle. For a `@Singleton` instance, that's when then `Injector` itself is destroyed, which is when the server shuts down.

Reply:
@ehumber, do we need to `awaitTermination`/`shutdownNow` here, to make in-flight produce responses are dispatched?

Reply:
Good spot Rigel :) fixed.